File: src/commands/scrollScreenshot.ts
import { invoke } from '@tauri-apps/api/core';

export enum ScrollDirection {
    /// 垂直滚动
    Vertical = 'Vertical',
    /// 水平滚动
    Horizontal = 'Horizontal',
}

export enum ScrollImageList {
    /// 上图片列表
    Top = 'Top',
    /// 下图片列表
    Bottom = 'Bottom',
}

export const scrollScreenshotInit = async (
    direction: ScrollDirection,
    imageWidth: number,
    imageHeight: number,
    sampleRate: number,
    minSampleSize: number,
    maxSampleSize: number,
    cornerThreshold: number,
    descriptorPatchSize: number,
    minSizeDelta: number,
    tryRollback: boolean,
) => {
    const result = await invoke('scroll_screenshot_init', {
        direction,
        imageWidth,
        imageHeight,
        sampleRate,
        minSampleSize,
        maxSampleSize,
        cornerThreshold,
        descriptorPatchSize,
        minSizeDelta,
        tryRollback,
    });
    return result;
};

export type ScrollScreenshotCaptureResult = {
    type: 'no_data' | 'no_change' | 'success' | 'no_image';
    thumbnail_buffer: ArrayBuffer | undefined;
    edge_position: number | undefined;
};

export const SCROLL_SCREENSHOT_CAPTURE_RESULT_EXTRA_DATA_SIZE = 4;

export const scrollScreenshotCapture = async (
    scrollImageList: ScrollImageList,
    monitorX: number,
    monitorY: number,
    minX: number,
    minY: number,
    maxX: number,
    maxY: number,
) => {
    const result = await invoke<ArrayBuffer>('scroll_screenshot_capture', {
        scrollImageList,
        monitorX,
        monitorY,
        minX,
        minY,
        maxX,
        maxY,
    });

    return result;
};

/**
 * @returns JPG 的 buffer 数据
 */
export const scrollScreenshotHandleImage = async (
    thumbnailSize: number,
): Promise<ScrollScreenshotCaptureResult> => {
    let result: ArrayBuffer | undefined;
    try {
        result = await invoke<ArrayBuffer>('scroll_screenshot_handle_image', {
            thumbnailSize,
        });
    } catch (error) {
        console.error('[scrollScreenshotHandleImage] error', error);
        result = new ArrayBuffer();
    }

    if (result.byteLength === 0) {
        return {
            type: 'no_data',
            thumbnail_buffer: result,
            edge_position: undefined,
        };
    }

    if (result.byteLength === 1) {
        const array = new Uint8Array(result);
        if (array[0] === 1) {
            return {
                type: 'no_change',
                thumbnail_buffer: undefined,
                edge_position: 0,
            };
        } else if (array[0] === 2) {
            return {
                type: 'no_image',
                thumbnail_buffer: undefined,
                edge_position: 0,
            };
        } else {
            return {
                type: 'no_data',
                thumbnail_buffer: new ArrayBuffer(),
                edge_position: undefined,
            };
        }
    }

    // 将屏幕信息和图像数据分离
    const imageDataLength = result.byteLength - SCROLL_SCREENSHOT_CAPTURE_RESULT_EXTRA_DATA_SIZE;

    // 提取屏幕信息
    const screenInfoView = new DataView(
        result,
        imageDataLength,
        SCROLL_SCREENSHOT_CAPTURE_RESULT_EXTRA_DATA_SIZE,
    );
    const edgePosition = screenInfoView.getInt32(0, true); // i32

    return {
        type: 'success',
        thumbnail_buffer: result,
        edge_position: edgePosition,
    };
};

export type ScrollScreenshotCaptureSize = {
    top_image_size: number;
    bottom_image_size: number;
};

export const scrollScreenshotGetSize = async () => {
    const result = await invoke<ScrollScreenshotCaptureSize>('scroll_screenshot_get_size');
    return result;
};

export const scrollScreenshotSaveToFile = async (filePath: string) => {
    const result = await invoke('scroll_screenshot_save_to_file', {
        filePath,
    });
    return result;
};

export const scrollScreenshotSaveToClipboard = async () => {
    const result = await invoke('scroll_screenshot_save_to_clipboard');
    return result;
};

export const scrollScreenshotClear = async () => {
    const result = await invoke('scroll_screenshot_clear');
    return result;
};

export const scrollScreenshotGetImageData = async (): Promise<Blob | undefined> => {
    let result: ArrayBuffer | undefined;
    try {
        result = await invoke<ArrayBuffer>('scroll_screenshot_get_image_data');
    } catch (error) {
        console.error(error);
        return undefined;
    }

    if (result.byteLength === 0) {
        return undefined;
    }

    return new Blob([result]);
};

File: src-tauri/src/scroll_screenshot.rs
use tauri::command;
use tauri::ipc::Response;
use tauri_plugin_clipboard_manager::ClipboardExt;
use tokio::sync::Mutex;

use snow_shot_app_scroll_screenshot_service::scroll_screenshot_image_service::ScrollScreenshotImageService;
use snow_shot_app_scroll_screenshot_service::scroll_screenshot_service::{
    ScrollDirection, ScrollImageList, ScrollScreenshotService,
};

#[command]
pub async fn scroll_screenshot_init(
    scroll_screenshot_service: tauri::State<'_, Mutex<ScrollScreenshotService>>,
    direction: ScrollDirection,
    sample_rate: f32,
    min_sample_size: u32,
    max_sample_size: u32,
    corner_threshold: u8,
    descriptor_patch_size: usize,
    min_size_delta: i32,
    try_rollback: bool,
) -> Result<(), ()> {
    snow_shot_tauri_commands_scroll_screenshot::scroll_screenshot_init(
        scroll_screenshot_service,
        direction,
        sample_rate,
        min_sample_size,
        max_sample_size,
        corner_threshold,
        descriptor_patch_size,
        min_size_delta,
        try_rollback,
    )
    .await
}

#[command]
pub async fn scroll_screenshot_capture(
    window: tauri::Window,
    scroll_screenshot_image_service: tauri::State<'_, Mutex<ScrollScreenshotImageService>>,
    scroll_image_list: ScrollImageList,
    monitor_x: i32,
    monitor_y: i32,
    min_x: u32,
    min_y: u32,
    max_x: u32,
    max_y: u32,
) -> Result<(), String> {
    snow_shot_tauri_commands_scroll_screenshot::scroll_screenshot_capture(
        window,
        scroll_screenshot_image_service,
        scroll_image_list,
        monitor_x,
        monitor_y,
        min_x,
        min_y,
        max_x,
        max_y,
    )
    .await
}

/**
 * 处理目前截取到的所有图片
 */
#[command]
pub async fn scroll_screenshot_handle_image(
    scroll_screenshot_service: tauri::State<'_, Mutex<ScrollScreenshotService>>,
    scroll_screenshot_image_service: tauri::State<'_, Mutex<ScrollScreenshotImageService>>,
    thumbnail_size: u32,
) -> Result<Response, ()> {
    snow_shot_tauri_commands_scroll_screenshot::scroll_screenshot_handle_image(
        scroll_screenshot_service,
        scroll_screenshot_image_service,
        thumbnail_size,
    )
    .await
}

#[command]
pub async fn scroll_screenshot_get_size(
    scroll_screenshot_service: tauri::State<'_, Mutex<ScrollScreenshotService>>,
) -> Result<snow_shot_tauri_commands_scroll_screenshot::ScrollScreenshotCaptureSize, ()> {
    snow_shot_tauri_commands_scroll_screenshot::scroll_screenshot_get_size(
        scroll_screenshot_service,
    )
    .await
}

#[command]
pub async fn scroll_screenshot_save_to_file(
    scroll_screenshot_service: tauri::State<'_, Mutex<ScrollScreenshotService>>,
    file_path: String,
) -> Result<(), String> {
    snow_shot_tauri_commands_scroll_screenshot::scroll_screenshot_save_to_file(
        scroll_screenshot_service,
        file_path,
    )
    .await
}

#[command]
pub async fn scroll_screenshot_save_to_clipboard(
    app: tauri::AppHandle,
    scroll_screenshot_service: tauri::State<'_, Mutex<ScrollScreenshotService>>,
) -> Result<(), String> {
    snow_shot_tauri_commands_scroll_screenshot::scroll_screenshot_save_to_clipboard(
        |image| match app.clipboard().write_image(&tauri::image::Image::new(
            image.as_bytes(),
            image.width(),
            image.height(),
        )) {
            Ok(_) => Ok(()),
            Err(e) => Err(format!(
                "[scroll_screenshot_save_to_clipboard] Failed to write image to clipboard: {}",
                e
            )),
        },
        scroll_screenshot_service,
    )
    .await
}

#[command]
pub async fn scroll_screenshot_clear(
    scroll_screenshot_service: tauri::State<'_, Mutex<ScrollScreenshotService>>,
) -> Result<(), ()> {
    snow_shot_tauri_commands_scroll_screenshot::scroll_screenshot_clear(scroll_screenshot_service)
        .await
}

#[command]
pub async fn scroll_screenshot_get_image_data(
    scroll_screenshot_service: tauri::State<'_, Mutex<ScrollScreenshotService>>,
) -> Result<Response, ()> {
    snow_shot_tauri_commands_scroll_screenshot::scroll_screenshot_get_image_data(
        scroll_screenshot_service,
    )
    .await
}
